### PR TITLE
[SPARK-6126][SQL] enforce UDTs as well in JsonRDDs

### PR DESCRIPTION
**Pull request by drubbo**

### CODE REVIEW

**Comment by rxin** [comment]
Is there a use case for this?

**Comment by drubbo** [comment]
This for instance: in [SparkGIS](https://github.com/drubbo/SparkGIS) we're adding support for geometry types that may be serialized in GeoJSON format, so that we'd like to do the following inside a Spark application:

    val schema = StructType(Seq(
        StructField("id", IntegerType),
        StructField("geo", GeometryType.Instance)
      ))

    val rdd = sparkContext.parallelize(Seq(
          "{\"id\":1,\"geo\":{\"type\":\"Point\",\"coordinates\":[1,1]}}",
          "{\"id\":2,\"geo\":{\"type\":\"LineString\",\"coordinates\":[[12,13],[15,20]]}}"
        ))

    val df = sqlContext.jsonRDD(rdd, schema) // here the schema is enforced

**Comment by rxin** [comment]
Thanks - can you add a unit test for this so we don't break it in the future?

**Comment by drubbo** [comment]
Sure. I'd also give a look at the comment of UserDefinedType.deserialize and make sure this usage of the method is reported as well.

**Comment by drubbo** [comment]
Should be done.

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by yhuai** [comment]
@drubbo We have added a new version of the JSON parser and we will remove the old one in future. Seems we handle UDTs in the new parser (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/json/JacksonParser.scala#L122-124)?

**Comment by drubbo** [comment]
Neat.
I saw the change in this JIRA, so this is expected to flow into spark-1.5 ?

**Comment by yhuai** [comment]
@drubbo Actually, if our new spark parser already handle UDTs, we can resolve the issue. Can you help us on that (confirming if the current spark 1.4 branch supports UDTs)?

**Comment by drubbo** [comment]
Absolutely, will do that asap.

**Comment by drubbo** [comment]
Everything works fine.
I would consider anyway adding the simple test and expanding the comment on deserialize as in my last two commits above https://github.com/drubbo/spark/commit/cd6d64d18154071e48a9dca443391dfffebcc7df and https://github.com/drubbo/spark/commit/e68ad407fa8aa821acbe185f62dd021eaa2e5331

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala [hunk 1]
```diff
@@ -53,7 +53,12 @@ abstract class UserDefinedType[UserType] extends DataType with Serializable {
    */
   def serialize(obj: Any): Any
 
-  /** Convert a SQL datum to the user type */
+  /**
+   * Convert a SQL datum to the user type
+   *
+   * This method may be called with an already deserialized datum, so it should be able to
+   * handle values of UserType too.
+   */
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: Where do we call it on a deserialized datum?

Comment (drubbo) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/UserDefinedType.scala]: I think it's related to https://issues.apache.org/jira/browse/SPARK-7186

**Comment by yhuai** [comment]
@drubbo Let's add the test in this PR to master. Can you create a new PR with your test and send a PR against master (also close this one)? Thanks!

**Pull request closed without merging**